ipam/allocate.go

```diff
@@ -42,9 +43,18 @@ func (g *allocate) Try(alloc *Allocator) bool {
 	}
 
 	// out of space
-	if donor, err := alloc.ring.ChoosePeerToAskForSpace(g.r.Start, g.r.End); err == nil {
+	tried := make(map[router.PeerName]struct{})
+	numPeersToAsk := alloc.ring.NumPeersToAskForSpace(g.r.Start, g.r.End)
+	for len(tried) < numPeersToAsk {
+		donor, _ := alloc.ring.ChoosePeerToAskForSpace(g.r.Start, g.r.End)
+		if _, triedAlready := tried[donor]; triedAlready {
+			continue
+		}
 		alloc.debugln("Decided to ask peer", donor, "for space in range", g.r)
-		alloc.sendSpaceRequest(donor, g.r)
+		if err := alloc.sendSpaceRequest(donor, g.r); err == nil {
+			break
+		}
+		tried[donor] = struct{}{}
```

Comment:
This has horrific complexity when there are lots of peers, most of which are dead.

I suggest we change the API s.t. `ChoosePeer*s*ToAskForSpace` returns a list in priority order.